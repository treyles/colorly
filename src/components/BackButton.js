import React from 'react';
import PropTypes from 'prop-types';
import Icon from '../utils/Icon';

const BackButton = ({ onClick }) => (
  <button className="back" onClick={onClick}>
    <Icon icon="back" />
    <span>BACK TO PALETTES</span>
  </button>
);

BackButton.defaultProps = {
  onClick: () => {}
};

BackButton.propTypes = {
  onClick: PropTypes.func
};

export default BackButton;
