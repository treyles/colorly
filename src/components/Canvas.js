import React from 'react';
import { storage } from '../utils/base';

export default class Canvas extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      mousePosition: {},
      imageLoaded: false,
      previewColor: false,
      currentColor: null,
      winWidth: null,
      winHeight: null
    };

    this.handleColorAndPosition = this.handleColorAndPosition.bind(this);
    this.handleCanvasLeave = this.handleCanvasLeave.bind(this);
    this.handleColorPick = this.handleColorPick.bind(this);
    this.updateWindowDimensions = this.updateWindowDimensions.bind(this);
  }

  componentDidMount() {
    this.ctx = this.canvas.getContext('2d');
    this.updateWindowDimensions();
    window.addEventListener('resize', this.updateWindowDimensions);
    this.getImageFile();
  }

  updateWindowDimensions() {
    this.setState({
      winWidth: window.innerWidth,
      winHeight: window.innerHeight
    });
  }

  // TODO: name handleColorPreview ???
  handleColorAndPosition(e) {
    const { imageLoaded } = this.state;

    const rect = this.canvas.getBoundingClientRect();
    const actualX = Math.floor(e.clientX - rect.left);
    const actualY = Math.floor(e.clientY - rect.top);

    const pixelData = this.ctx.getImageData(actualX, actualY, 1, 1);
    const { data } = pixelData;
    const pixelColor = `rgba(${data[0]}, ${data[1]}, ${data[2]}, ${data[3]})`;

    if (imageLoaded) {
      this.setState({
        currentColor: pixelColor,
        mousePosition: { x: actualX, y: actualY },
        previewColor: true
      });
    }
  }

  handleCanvasLeave() {
    this.setState({
      previewColor: false
    });
  }

  handleColorPick() {
    const { currentColor } = this.state;
    this.props.makePalette(currentColor);
  }

  drawCanvasImage(image) {
    const heightPadding = 275;
    const widthPadding = 60;
    const containerWidth = this.state.winWidth - widthPadding;
    const containerHeight = this.state.winHeight - heightPadding;
    let imageWidth = image.width;
    let imageHeight = image.height;

    // if any part of the image (original dimensions) overflow
    // the container, shrink dimensions and maintain aspect ratio
    if (imageWidth > containerWidth || imageHeight > containerHeight) {
      // algorithm modified from 1owk3y: https://bit.ly/2JiYzHz
      const wFits = containerWidth / imageWidth;
      const hFits = containerHeight / imageHeight;
      const minFits = wFits > hFits ? hFits : wFits;

      imageWidth *= minFits;
      imageHeight *= minFits;
    }

    this.canvas.width = imageWidth;
    this.canvas.height = imageHeight;

    // TODO: check if this works
    this.ctx.webkitImageSmoothingEnabled = false;
    this.ctx.drawImage(image, 0, 0, imageWidth, imageHeight);
  }

  getImageFile() {
    const imageFile = storage
      .ref()
      .child(`images/${this.props.imageName}`);

    imageFile.getDownloadURL().then(url => {
      const image = new Image();
      image.crossOrigin = 'anonymous';

      image.onload = () => {
        this.drawCanvasImage(image);
        this.setState({ imageLoaded: true });
      };

      image.src = url;
    });
  }

  render() {
    const { mousePosition, previewColor, currentColor } = this.state;

    const containerStyle = {
      top: `${mousePosition.y}px`,
      left: `${mousePosition.x + 10}px`
    };

    const colorStyle = {
      background: currentColor
    };

    return (
      <div className="canvas-wrapper">
        <canvas
          ref={ref => {
            this.canvas = ref;
          }}
          // width={800}
          // height={550}
          onMouseMove={this.handleColorAndPosition}
          onMouseLeave={this.handleCanvasLeave}
          onClick={this.handleColorPick}
        />
        {previewColor && (
          <div className="preview-container" style={containerStyle}>
            <div className="preview-color" style={colorStyle} />
          </div>
        )}
      </div>
    );
  }
}
