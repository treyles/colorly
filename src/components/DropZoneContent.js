import React from 'react';
import PropTypes from 'prop-types';

export default class DropZoneContent extends React.Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange() {
    this.props.handleImageDrop();
  }

  render() {
    const { imageAlert } = this.props;

    return (
      <div className="drop-zone-content">
        <label htmlFor="file-browse" className={imageAlert && 'alert'}>
          {imageAlert ? (
            <h1>{imageAlert}</h1>
          ) : (
            <div>
              <h1>Drag and Drop Image Here</h1>
              <h3>JPG, GIF, PNG or SVG no larger than 3 MB</h3>
            </div>
          )}
        </label>
        <input
          id="file-browse"
          type="file"
          accept="image/*"
          onChange={this.handleChange}
        />
      </div>
    );
  }
}

DropZoneContent.defaultProps = {
  imageAlert: null
};

DropZoneContent.propTypes = {
  imageAlert: PropTypes.string,
  handleImageDrop: PropTypes.func.isRequired
};
