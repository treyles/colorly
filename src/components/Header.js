/* eslint-disable */
import React from 'react';
import Icon from '../utils/Icon';
import { Link } from 'react-router-dom';
import firebase from 'firebase';
import ClickOutside from '../utils/ClickOutside';
// import Icon from '../utils/Icon';

export default class Header extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      profileOpen: false
    };

    this.handleSignOutClick = this.handleSignOutClick.bind(this);
    this.handleDialogToggleClick = this.handleDialogToggleClick.bind(this);
    this.handleAddPaletteClick = this.handleAddPaletteClick.bind(this);
    this.closeDialog = this.closeDialog.bind(this);
  }

  handleSignOutClick() {
    firebase
      .auth()
      .signOut()
      .then(() => {
        console.log('signed out');
        localStorage.setItem('authenticated', false);
      });
  }

  handleDialogToggleClick() {
    this.setState({
      profileOpen: !this.state.profileOpen
    });
  }

  handleAddPaletteClick() {
    if (this.props.isNewUser) {
      this.props.closeNewUserDialog();
    }

    this.props.history.push('/upload');
  }

  closeDialog() {
    this.setState({
      profileOpen: false
    });
  }

  render() {
    const { profileOpen } = this.state;
    const { currentUser, library, animateHeader } = this.props;

    const profileDialog = (
      <div className="profile-dialog">
        <div className="palette-count">
          <span className="count">{library.length}</span>
          {library.length !== 1 ? 'PALETTES' : 'PALETTE'}
        </div>
        <button className="sign-out" onClick={this.handleSignOutClick}>
          SIGN OUT
        </button>
      </div>
    );

    return (
      <header>
        <div
          className="header-top"
          style={{ backgroundColor: `${animateHeader ? '#fff' : ''}` }}
        >
          <div className="profile">
            <div className="profile-image">
              {currentUser.photoURL && <img src={currentUser.photoURL} />}
            </div>
            <span onClick={this.handleDialogToggleClick}>
              {currentUser.displayName}
            </span>
            <span onClick={this.handleDialogToggleClick}>
              <Icon icon="down" />
            </span>
            <ClickOutside
              elementIsOpen={profileOpen}
              onRequestClose={this.closeDialog}
            >
              {profileDialog}
            </ClickOutside>
          </div>
          {/* <Link to="/upload"> */}
          <button className="add" onClick={this.handleAddPaletteClick}>
            {/* <span>
                <Icon icon="add" />
              </span> */}
            ADD PALETTE
          </button>
          {/* </Link> */}
        </div>
        <div
          className="header-logo"
          style={{
            opacity: `${animateHeader ? 0 : ''}`,
            visibility: `${animateHeader ? 'hidden' : 'visible'}`
          }}
        >
          <h2>Colorly</h2>
          <span>
            <Icon icon="logo" />
          </span>
        </div>
      </header>
    );
  }
}

// style={{ background: `${animateHeader ? '#fff' : ''}` }}
// style={{ opacity: `${animateHeader ? 0 : ''}` }}
