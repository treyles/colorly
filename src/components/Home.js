import React from 'react';
import PropTypes from 'prop-types';
import Icon from '../utils/Icon';
import { googleAuth } from '../utils/base';

export default function Home({ signInUser }) {
  return (
    <div className="home">
      <div className="hero-text-container">
        <div className="content">
          <h1>Colorly</h1>
          <span className="logo">
            <Icon icon="logo" />
          </span>
          <h3 className="pitch">
            A simple color extraction tool. Save custom palettes to your
            collection along with the source image for future reference!
          </h3>
          <div className="buttons">
            <button className="twitter-login">Connect with Twitter</button>
            <button
              className="google-login"
              onClick={() => signInUser(googleAuth)}
            >
              Connect with Google
            </button>
          </div>
        </div>
      </div>
      <div className="hero-image" />
    </div>
  );
}

Home.propTypes = {
  signInUser: PropTypes.func.isRequired
};
