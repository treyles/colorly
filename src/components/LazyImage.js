import React from 'react';
// import PropTypes from 'prop-types'

export default class LazyImage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      imageLoaded: false
    };

    this.handleLoad = this.handleLoad.bind(this);
  }

  componentDidMount() {
    this.img = new Image();
    this.img.onload = this.handleLoad;
    this.img.src = this.props.url;
  }

  handleLoad() {
    this.setState({
      imageLoaded: true
    });
  }

  render() {
    const preloader = (
      <div className="preloader-wrapper">
        <span className="preloader">Loading</span>
      </div>
    );

    return this.state.imageLoaded ? (
      <img className="image" src={this.props.url} alt="" />
    ) : (
      preloader
    );
  }
}
