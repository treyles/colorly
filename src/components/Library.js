import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { demoRef, databaseRef } from '../utils/base';
import Header from './Header';
import Icon from '../utils/Icon';
import PaletteCard from './PaletteCard';
import BackButton from './BackButton';
import LazyImage from '../utils/LazyImage';
import { closeNewUserDialog, fetchLibrary } from '../actions';

export class Library extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      imageSource: null,
      animateHeader: false
    };

    this.setImageSource = this.setImageSource.bind(this);
    this.handleCloseImageClick = this.handleCloseImageClick.bind(this);
    this.scrollPosition = this.scrollPosition.bind(this);
    this.addDemoPalettes = this.addDemoPalettes.bind(this);
  }

  componentDidMount() {
    window.addEventListener('scroll', this.scrollPosition);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.scrollPosition);
  }

  setImageSource(url) {
    this.setState({
      imageSource: url
    });
  }

  handleCloseImageClick() {
    this.setState({
      imageSource: null
    });
  }

  addDemoPalettes() {
    const { uid, closeNewUserDialog } = this.props;
    demoRef
      .once('value')
      .then(snapshot => snapshot.val())
      .then(palettes => {
        databaseRef.child(uid).set(palettes);
      });

    closeNewUserDialog();
  }

  scrollPosition() {
    let { animateHeader } = this.state;

    if (window.scrollY > 40) {
      animateHeader = true;
    } else {
      animateHeader = false;
    }

    this.setState({ animateHeader });
  }

  render() {
    const { imageSource, animateHeader } = this.state;
    const { library, loading, isNewUser } = this.props;

    const placeholders = [...Array(10)].map((_, index) => (
      <div key={index} className="placeholder" />
    ));

    const viewImage = (
      <div className="image-view">
        <BackButton onClick={this.handleCloseImageClick} />
        <LazyImage url={imageSource} />
      </div>
    );

    const emptyMessage = (
      <div className="empty-library">
        <h3>Empty Palette Collection!</h3>
      </div>
    );

    const newUserDialog = (
      <div className="new-user">
        <Icon icon="handWave" />
        <h3>Hey! It Looks Like Youâ€™re New</h3>
        <h4>
          To get started making color palettes, load an image by clicking
          the Add Palette button on the top right corner.
        </h4>
        <button
          className="dismiss-btn"
          onClick={() => this.props.closeNewUserDialog()}
        >
          DISMISS
        </button>
        <button
          className="lazy-btn"
          onClick={() => this.addDemoPalettes()}
        >
          FEELING LAZY?
        </button>
      </div>
    );

    return (
      <div className="library">
        <Header animateHeader={animateHeader} />
        {library.map(palette => (
          <PaletteCard
            key={palette.id}
            data={palette}
            setImageSource={this.setImageSource}
          />
        ))}
        {loading && !isNewUser && placeholders}
        {imageSource && viewImage}
        {!library.length && !loading && emptyMessage}
        {isNewUser && newUserDialog}
      </div>
    );
  }
}

Library.defaultProps = {
  uid: null
};

Library.propTypes = {
  uid: PropTypes.string,
  library: PropTypes.arrayOf(PropTypes.object).isRequired,
  loading: PropTypes.bool.isRequired,
  isNewUser: PropTypes.bool.isRequired,
  closeNewUserDialog: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  library: state.data.library,
  loading: state.data.loading,
  isNewUser: state.user.isNewUser,
  uid: state.user.currentUser.uid
});

export default connect(mapStateToProps, {
  closeNewUserDialog,
  fetchLibrary
})(Library);
