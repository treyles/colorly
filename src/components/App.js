import React from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import Upload from './Upload';
import Library from './Library';

const testData = [
  {
    url:
      'https://firebasestorage.googleapis.com/v0/b/colorly-44942.appspot.com/o/images%2Fa8140b0e-748a-4a8d-a9fe-91324f4174c5?alt=media&token=5ca4aef7-31cd-4389-b5f5-45da4292f057',
    palette: {
      color1: '#37BEB7'
    },
    title: 'Single Color',
    id: 'a8140b0e-748a-4a8d-a9fe-91324f4174c5'
  },
  {
    url:
      'https://firebasestorage.googleapis.com/v0/b/colorly-44942.appspot.com/o/images%2F2ec38480-3daf-4c9e-8a68-f2d9de46f6cb?alt=media&token=a8e97814-1650-48d7-9860-690fc936ecf5',
    palette: {
      color1: '#126794',
      color2: '#F8AC8E',
      color3: '#EF6868',
      color4: '#F2C61B',
      color5: '#DDDDDD'
    },
    title: 'Abstract Drawing',
    id: '2ec38480-3daf-4c9e-8a68-f2d9de46f6cb'
  },
  {
    url:
      'https://firebasestorage.googleapis.com/v0/b/colorly-44942.appspot.com/o/images%2F0917631b-26f3-46a3-804c-f5ecc1355e39?alt=media&token=d0e2052a-d3f8-4fbc-b430-b7ae259ed9f3',
    palette: {
      color1: '#49124B',
      color2: '#AE53A0',
      color3: '#F5966A',
      color4: '#F04267'
    },
    title: 'Uber Illustration Style',
    id: '0917631b-26f3-46a3-804c-f5ecc1355e39'
  }
];

export default class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      library: testData
    };

    this.addPaletteToLibrary = this.addPaletteToLibrary.bind(this);
  }

  addPaletteToLibrary(obj) {
    const { library } = this.state;
    this.setState({
      library: [obj].concat(library)
    });
  }

  render() {
    const { library } = this.state;

    return (
      <BrowserRouter>
        <Switch>
          <Route
            exact
            path="/"
            render={() => <Library library={library} />}
          />
          <Route
            path="/upload"
            render={({ history }) => (
              <Upload
                history={history}
                addPaletteToLibrary={this.addPaletteToLibrary}
              />
            )}
          />
        </Switch>
      </BrowserRouter>
    );
  }
}

// <div>
//   <Header />
// <Upload addPaletteToLibrary={this.addPaletteToLibrary} />
// </div>
