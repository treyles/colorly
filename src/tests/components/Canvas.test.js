import React from 'react';
import { shallow } from 'enzyme';
import { Canvas } from '../../components/Canvas';

describe('Canvas', () => {
  const props = {
    addColorToPalette: jest.fn(),
    imageSource: {}
  };

  beforeAll(() => {
    Canvas.prototype.componentDidMount = null;
  });

  it('renders correctly', () => {
    const component = shallow(<Canvas {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('sets winWidth and winHeight state on browser resize', () => {
    const component = shallow(<Canvas {...props} />);
    global.innerWidth = 444;
    global.innerHeight = 333;
    component.instance().updateBrowserSize();
    expect(component.state().winWidth).toEqual(444);
    expect(component.state().winHeight).toEqual(333);
  });

  it('returns optimized width and height based on window dimensions', () => {
    const component = shallow(<Canvas {...props} />);
    component.setState({
      winWidth: 800,
      winHeight: 600
    });
    // mock image
    const initialImageSize = {
      width: 1024,
      height: 800
    };
    const result = component.instance().optimizeScale(initialImageSize);
    expect(result.width < 800).toEqual(true);
    expect(result.height < 600).toEqual(true);
  });

  it('calls "handleColorPreviewMove" on canvas move', () => {
    const spy = jest
      .spyOn(Canvas.prototype, 'handleColorPreviewMove')
      .mockImplementation(jest.fn());
    const component = shallow(<Canvas {...props} />);
    component.find('canvas').simulate('mousemove');
    expect(spy).toHaveBeenCalled();
    spy.mockClear();
  });

  it('calls "handleCanvasLeave" on canvas leave', () => {
    const spy = jest.spyOn(Canvas.prototype, 'handleCanvasLeave');
    const component = shallow(<Canvas {...props} />);
    component.find('canvas').simulate('mouseleave');
    expect(spy).toHaveBeenCalled();
    spy.mockClear();
  });

  it('calls "handleColorClick" on canvas click', () => {
    const spy = jest.spyOn(Canvas.prototype, 'handleColorClick');
    const component = shallow(<Canvas {...props} />);
    component.find('canvas').simulate('click');
    expect(spy).toHaveBeenCalled();
    spy.mockClear();
  });
});
