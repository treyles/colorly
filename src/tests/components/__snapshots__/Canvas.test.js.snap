// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Canvas
    imageSource={Object {}}
    makePalette={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <canvas
          onClick={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
        />,
        false,
      ],
      "className": "canvas-wrapper",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onClick": [Function],
          "onMouseLeave": [Function],
          "onMouseMove": [Function],
        },
        "ref": [Function],
        "rendered": null,
        "type": "canvas",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <canvas
            onClick={[Function]}
            onMouseLeave={[Function]}
            onMouseMove={[Function]}
          />,
          false,
        ],
        "className": "canvas-wrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onClick": [Function],
            "onMouseLeave": [Function],
            "onMouseMove": [Function],
          },
          "ref": [Function],
          "rendered": null,
          "type": "canvas",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
